File: Creational/Lab6/C#/EstatePool.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace C_
{
    public class EstatePool
    {
        private static Stack<RealEstate> _pool = new Stack<RealEstate>();
        private static object _lock = new object();

        public static RealEstate GetInstance(string type)
        {
            lock(_lock){            
                if (_pool.Count > 0)
                {
                    foreach(var v in _pool){
                        if(v.GetType().Name.ToLower() == type){
                            Console.WriteLine($"{type} with hash {v.GetHashCode()} was returned.");
                            return _pool.Pop();
                        }
                    }
                    
                }
                var instance = EstateFactory(type);
                // Console.WriteLine($"{type} with hash {instance.GetHashCode()} was created.");
                return instance;
            }
        }

        private static RealEstate EstateFactory(string type) => type.ToLower() switch 
        {
            "apartment" => new Apartment(),
            "room" => new Room(),
            "house" => new House(),
            _ => throw new Exception("Wrong type of estate")
        };
        

        public static void ReleaseInstance(RealEstate instance)
        {
            lock(_lock){ 
                instance.Reset();
                _pool.Push(instance);
            }
        }
    }
}

File: Creational/Lab6/Python/estate_pool.py
from threading import Lock
from typing import List
from real_estate import *
from apartment import *
from room import *
from house import *
import sys

class EstatePool:
    _pool = []
    _lock = Lock()

    @staticmethod
    def get_instance(type_: str):
        with EstatePool._lock:
            for i, estate in enumerate(EstatePool._pool):
                if estate.__class__.__name__.lower() == type_.lower():
                    print(f"{type_} with hash {hash(estate)} was returned.")
                    return EstatePool._pool.pop(i)
            
            instance = EstatePool.estate_factory(type_)
            # print(f"{type_} with hash {hash(instance)} was created.")
            return instance

    @staticmethod
    def estate_factory(type_: str):
        type_ = type_.lower()
        if type_ == "apartment":
            return Apartment()
        elif type_ == "room":
            return Room()
        elif type_ == "house":
            return House()
        else:
            raise Exception("Wrong type of estate")

    @staticmethod
    def release_instance(instance):
        with EstatePool._lock:
            instance.reset()
            EstatePool._pool.append(instance)